File: benchmarks/loop_sum.f
// A polynomial-hash-style accumulation, not a plain running sum -- a
// plain `total = total + i` loop is a well-known pattern optimizing
// compilers (LLVM in particular) can collapse into a closed-form
// expression at compile time, which would benchmark the optimizer's
// algebra rather than the generated loop's actual runtime cost (verified:
// a plain sum loop up to 200,000,000 here ran in ~2ms, meaning it had
// already been folded away entirely). Each iteration genuinely depends
// on the previous one (modular multiply-add), so there's no closed form
// to fold it into -- every language here has to actually run the loop.
int total = 0
for int i = 0, i < 100000000, i++ {
    total = (total * 1000003 + i) % 1000000007
}
log(total)

File: benchmarks/loop_sum.c
#include <stdio.h>
#include <stdint.h>

/* See loop_sum.f's own comment: a polynomial-hash-style accumulation,
 * not a plain running sum -- a plain sum is exactly the kind of loop an
 * optimizing compiler can fold into a closed-form expression at compile
 * time. Every iteration here genuinely depends on the previous one. */
int main(void) {
    int64_t total = 0;
    for (int64_t i = 0; i < 100000000; i++) {
        total = (total * 1000003 + i) % 1000000007;
    }
    printf("%lld\n", (long long)total);
    return 0;
}
